src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -2,22 +2,130 @@
+        // recompute and cache with the new state sets.
+        if (!s_projectToCompilationWithAnalyzers.TryGetValue(project, out var tupleBox) ||
+            !stateSets.IsSubsetOf(tupleBox.Value.stateSets))
+        {
+            var compilationWithAnalyzersPair = await CreateCompilationWithAnalyzersAsync().ConfigureAwait(false);
+            tupleBox = new((stateSets, compilationWithAnalyzersPair));
+
+            // Make a best effort attempt to store the latest computed value against these state sets. If this
+            // fails (because another thread interleaves with this), that's ok.  We still return the pair we 
+            // computed, so our caller will still see the right data
+            s_projectToCompilationWithAnalyzers.Remove(project);
+
+            // Intentionally ignore the result of this.  We still want to use the value we computed above, even if
+            // another thread interleaves and sets a different value.
+            s_projectToCompilationWithAnalyzers.GetValue(project, _ => tupleBox);
+        }
+
+        return tupleBox.Value.compilationWithAnalyzersPair;
+
+        // <summary>
+        // Should only be called on a <see cref="Project"/> that <see cref="Project.SupportsCompilation"/>.
+        // </summary>
+        async Task<CompilationWithAnalyzersPair?> CreateCompilationWithAnalyzersAsync()
+        {
+            var projectAnalyzers = stateSets.SelectAsArray(s => !s.IsHostAnalyzer, s => s.Analyzer);
+            var hostAnalyzers = stateSets.SelectAsArray(s => s.IsHostAnalyzer, s => s.Analyzer);
+
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            // Create driver that holds onto compilation and associated analyzers
+            var filteredProjectAnalyzers = projectAnalyzers.WhereAsArray(static a => !a.IsWorkspaceDiagnosticAnalyzer());
+            var filteredHostAnalyzers = hostAnalyzers.WhereAsArray(static a => !a.IsWorkspaceDiagnosticAnalyzer());
+            var filteredProjectSuppressors = filteredProjectAnalyzers.WhereAsArray(static a => a is DiagnosticSuppressor);
+            filteredHostAnalyzers = filteredHostAnalyzers.AddRange(filteredProjectSuppressors);
+
+            // PERF: there is no analyzers for this compilation.
+            //       compilationWithAnalyzer will throw if it is created with no analyzers which is perf optimization.
+            if (filteredProjectAnalyzers.IsEmpty && filteredHostAnalyzers.IsEmpty)
+            {
+                return null;
+            }
+
+            Contract.ThrowIfFalse(project.SupportsCompilation);
+            AssertCompilation(project, compilation);
+
+            // in IDE, we always set concurrentAnalysis == false otherwise, we can get into thread starvation due to
+            // async being used with synchronous blocking concurrency.
+            var projectAnalyzerOptions = new CompilationWithAnalyzersOptions(
```

Comment:
Would it be worthwhile to only construct the options if we knew we were going to use them?

Reply:
yes.  that would make sense.  this logic exists from before.  but i can make that change later!

Reply:
Made in 3c5c6868871